### ISSUE TITLE
Docker build should compute image digests

### ISSUE THREAD

**Issue by phs** [docker/docker#32016]
**Description**

Currently `docker build` does not compute and save the digests of images that it creates.  Instead, images must be pushed to a registry for their digests to be assigned.

This creates headaches when using `docker stack` when no registry is present, such in local development with a single-node swarm.

Some related discussion:
* https://groups.google.com/d/msg/docker-user/PvAcxDrvP30/ZP7uUwWO084J
> You don't see a digest when doing 'docker build' because the registry is currently responsible for calculating the digest of the uploaded image manifest and returning that value to the client. The reason for this is that we don't have a guaranteed way for every client and registry implementation to consistently calculate the same digest for the same input. There is a specification for the canonical JSON format that would in theory allow the daemon to calculate the digest, but it has not been implemented yet: https://github.com/docker/distribution/blob/master/docs/spec/json.md.

That was 2015.  Perhaps there has been some motion since?

* https://github.com/docker/docker/pull/28173 https://github.com/docker/docker/issues/29205 https://github.com/docker/docker/issues/31357 https://github.com/docker/docker/issues/31518

Swarm mode begins use digests to track images, along with some example headaches introduced by missing digests.

**Steps to reproduce the issue:**

```
phs:~/bug $ cat Dockerfile
FROM alpine
CMD echo Create a custom image
phs:~/bug $ docker build -t myimage:latest .
Sending build context to Docker daemon 2.048 kB
Step 1/2 : FROM alpine
 ---> 88e169ea8f46
Step 2/2 : CMD echo Create a custom image
 ---> Running in 0cd8b56d7e58
 ---> 6547a9192706
Removing intermediate container 0cd8b56d7e58
Successfully built 6547a9192706
phs:~/bug $ docker image ls --digests myimage:latest
REPOSITORY          TAG                 DIGEST              IMAGE ID            CREATED              SIZE
myimage             latest              <none>              6547a9192706        About a minute ago   3.98 MB
```

**Describe the results you received:**

> `<none>`

**Describe the results you expected:**

An image digest computed and saved without any need to move the image somewhere else.

**Output of `docker version`:**

```
phs:~/bug $ docker version
Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 10:40:59 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   3a232c8
 Built:        Tue Feb 28 07:52:04 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
phs:~/bug $ docker info
Containers: 4
 Running: 4
 Paused: 0
 Stopped: 0
Images: 89
Server Version: 17.03.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: active
 NodeID: 4hty7de0lmnsyqjqweuc30mki
 Is Manager: true
 ClusterID: e7x2iyfpa13c584j7euo1fg6e
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 192.168.65.2
 Manager Addresses:
  192.168.65.2:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 977c511eda0925a723debdc94d09459af49d082a
runc version: a01dafd48bc1c7cc12bdb01206f9fea7dd6feb70
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.12-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.952 GiB
Name: moby
ID: MHAO:ZBB5:PYQW:2LSF:T7KL:3KQJ:5TUB:GLQQ:DRI6:WUGB:ES4T:ADBR
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 171
 Goroutines: 340
 System Time: 2017-03-22T21:08:49.774201311Z
 EventsListeners: 5
No Proxy: *.local, 169.254/16
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

@thaJeztah @nishanttotla

**Comment by nishanttotla** [comment]
@phs perhaps https://github.com/docker/docker/issues/22011#issuecomment-213515253 answers your question.

Could you specify what your stack file/service create commands look like, so that we can try to fix the specific issue you're facing?

**Comment by phs** [comment]
Thanks, that highlights the trade-off well:

> We've explored compressing layers when they are first created to make it possible for manifests to be persistent like you suggest, but there are two issues with doing this:
>
> * docker build would become a lot slower, even if you never push the resulting image to a registry.
> * Images would use up to twice as much disk space, because the compressed version of each layer has to be stored as well as the usable format of the layer.

I would support an opt-in option (`docker build --digest`?) that pays the extra time and eagerly computes the digests.

I'm not sure I follow the 2nd point.  Computing the digests locally shouldn't use any more disk than the situation where the image is received from a registry.  When a registry is present, either the client saves the compressed form or it doesn't (and trusts the registry about the validity of the digest).  The registry-less scenario can do the same thing (compute the digest and if desired throw the compressed form away).

**Comment by phs** [comment]
You asked what my specific configuration was.  My situation is similar to #31518.

Particularly, I'm trying to migrate away from `docker-compose` and create a developer workflow based on `docker stack`.  My build script now explicitly builds local images and runs them in a single-node swarm.  There is no registry.  At the moment I am considering how to best update the services in my local swarm after a new image is built.  This has led me to bump into problems similar to the issues I linked above.

Using `--force` with `docker service update` works.  It's just not a very clean solution.  For one thing, it requires me to dereference the service ID to issue the update:

```
for i in $(docker stack services -q --filter name="$service" "$stack")
do
  docker service update --force $i
done
```

If we want to update just those services running stale images, we need to jump through more hoops to identify them.

More generally it indicates a feature gap for using swarm for isolated local development, which seems like a natural and popular use case.  A robust way to either exhibit or not require image digests would attack the problem at the root.

**Comment by jharveysmith** [comment]
Has there been any progress on this?

**Comment by phs** [comment]
I've long since moved on [building images with bazel][0] (and deploying them on kubernetes.)

[0]: https://github.com/bazelbuild/rules_docker

**Comment by nustiueudinastea** [comment]
Bazel is a pretty complex solution. It would be fantastic if this would be added to Moby.

**Comment by thaJeztah** [comment]
ping @tonistiigi @AkihiroSuda @dmcgowan is this something that can be implemented with the containerd integration?

**Comment by AkihiroSuda** [comment]
iidfile works?

**Comment by wallrj** [comment]
> iidfile works?

@AkihiroSuda No, the `docker build --iidfile` content is the same as the sha256 image ID which is printed by `docker build`.

But this is not the same as the `docker images --digests` value, which I understand to be a hash that is currently only calculated by the Docker Registry server, and only becomes available to the local docker daemon when the image is pulled from a Docker registry.

My use-case is that I'm using `eval $(minikube docker-env)` so that images are built by the docker daemon running in a minikube VM.
I then want to be able to run Kubernetes Pods docker image using an `<image-name>@<digest>` syntax.
But this doesn't work, because the docker daemon  inside minikube doesn't know the digest.

I like @phs 's `docker build --digest` idea (above), if that's possible.

**Comment by wallrj** [comment]
> iidfile works?

@AkihiroSuda No, the `docker build --iidfile` content is the same as the sha256 image ID which is printed by `docker build`.

But this is not the same as the `docker images --digests` value, which I understand to be a hash that is currently only calculated by the Docker Registry server, and only becomes available to the local docker daemon when the image is pulled from a Docker registry.

My use-case is that I'm using `eval $(minikube docker-env)` so that images are built by the docker daemon running in a minikube VM.
I then want to be able to run Kubernetes Pods docker image using an `<image-name>@<digest>` syntax.
But this doesn't work, because the docker daemon  inside minikube doesn't know the digest.

I like @phs 's `docker build --digest` idea (above), if that's possible.
